File: go/internal/monitor/http_client.go
package monitor

import (
	"io"
	"net/http"
	"regexp"
	"strings"
	"time"
)

func newHTTPClient(timeoutSeconds int) *http.Client {
	if timeoutSeconds < 5 {
		timeoutSeconds = 5
	}
	return &http.Client{Timeout: time.Duration(timeoutSeconds) * time.Second}
}

func firstTitleFromHTML(html string) string {
	if len(html) > 250000 {
		html = html[:250000]
	}
	re := regexp.MustCompile(`(?is)<title[^>]*>(.*?)</title>`)
	match := re.FindStringSubmatch(html)
	if len(match) < 2 {
		return ""
	}
	return strings.Join(strings.Fields(strings.TrimSpace(match[1])), " ")
}

func performRequest(client *http.Client, method string, rawURL string, headers map[string]string, body io.Reader) (*http.Response, []byte, error) {
	req, err := http.NewRequest(method, rawURL, body)
	if err != nil {
		return nil, nil, newFetchError("build request failed for %s: %v", rawURL, err)
	}
	req.Header.Set("User-Agent", "CareerPageMonitor/3.0 (+go-swarm)")
	req.Header.Set("Accept", "application/json,text/html;q=0.9,*/*;q=0.8")
	for key, value := range headers {
		req.Header.Set(key, value)
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, nil, newFetchError("request failed for %s: %v", rawURL, err)
	}
	defer resp.Body.Close()
	payload, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, nil, newFetchError("read response failed for %s: %v", rawURL, err)
	}

	title := ""
	contentType := strings.ToLower(strings.TrimSpace(resp.Header.Get("Content-Type")))
	if strings.Contains(contentType, "text/html") {
		title = firstTitleFromHTML(string(payload))
	}
	if classifyHTTPBlock(resp.StatusCode, title) {
		detail := "Blocked with HTTP " + resp.Status
		if title != "" {
			detail += " (" + title + ")"
		}
		return resp, payload, &CrawlBlockedError{Message: detail}
	}
	if resp.StatusCode >= 400 {
		return resp, payload, &CrawlFetchError{Message: "HTTP " + resp.Status + " for " + rawURL}
	}
	return resp, payload, nil
}

File: career_monitor/http_client.py
from __future__ import annotations

from typing import Any

import requests
from bs4 import BeautifulSoup

from .exceptions import CrawlBlockedError, CrawlFetchError
from .utils import classify_http_block


def requests_session() -> requests.Session:
    session = requests.Session()
    session.headers.update(
        {
            "User-Agent": "CareerPageMonitor/2.1 (+crawler-swarm)",
            "Accept": "application/json,text/html;q=0.9,*/*;q=0.8",
        }
    )
    return session


def first_title_from_html(html: str) -> str:
    soup = BeautifulSoup(html[:250000], "html.parser")
    if soup.title and soup.title.string:
        return " ".join(soup.title.string.split()).strip()
    return ""


def perform_request(
    session: requests.Session,
    method: str,
    url: str,
    timeout_seconds: int,
    **kwargs: Any,
) -> requests.Response:
    response = session.request(method=method, url=url, timeout=timeout_seconds, **kwargs)
    content_type = (response.headers.get("content-type") or "").lower()
    title = ""
    if "text/html" in content_type:
        title = first_title_from_html(response.text)
    if classify_http_block(response.status_code, title):
        detail = f"Blocked with HTTP {response.status_code}"
        if title:
            detail = f"{detail} ({title})"
        raise CrawlBlockedError(detail)
    if response.status_code >= 400:
        raise CrawlFetchError(f"HTTP {response.status_code} for {url}")
    return response

